src/util/antlr/ANTLRErrorHandling.cpp

```diff
@@ -2,38 +2,76 @@
 //  Chair of Algorithms and Data Structures.
 //  Author: Julian Mundhahs (mundhahj@informatik.uni-freiburg.de)
 
-#include <util/antlr/ANTLRErrorHandling.h>
+#include "ANTLRErrorHandling.h"
 
-#include "CharStream.h"
-#include "CommonTokenStream.h"
-#include "TokenSource.h"
+#include <Lexer.h>
+#include <Parser.h>
+
+#include "util/Exception.h"
+
+// _____________________________________________________________________________
+ExceptionMetadata generateMetadata(antlr4::Parser* parser,
+                                   antlr4::Token* offendingToken, size_t line,
+                                   size_t charPositionInLine) {
+  return {
+      parser->getTokenStream()->getTokenSource()->getInputStream()->toString(),
+      offendingToken->getStartIndex(), offendingToken->getStopIndex(), line,
+      charPositionInLine};
+}
+
+// _____________________________________________________________________________
+ExceptionMetadata generateMetadata(antlr4::Lexer* lexer, size_t line,
+                                   size_t charPositionInLine) {
+  return {lexer->getInputStream()->toString(), lexer->tokenStartCharIndex,
+          lexer->tokenStartCharIndex, line, charPositionInLine};
+}
 
 // _____________________________________________________________________________
 ExceptionMetadata generateMetadata(antlr4::Recognizer* recognizer,
-                                   antlr4::Token* offendingToken,
-                                   [[maybe_unused]] size_t line,
-                                   [[maybe_unused]] size_t charPositionInLine) {
-  return {((antlr4::CommonTokenStream*)recognizer->getInputStream())
-              ->getTokenSource()
-              ->getInputStream()
-              ->toString(),
-          offendingToken->getStartIndex(), offendingToken->getStopIndex()};
+                                   antlr4::Token* offendingToken, size_t line,
+                                   size_t charPositionInLine) {
+  // The `antlr4::Recognizer` interface does not provide enough information to
+  // obtain a copy of the input string. Do this switching here to obtain that.
+  // As of ANTLRv4 `antlr4::Parser` and `antlr4::Lexer` are the only subclasses
+  // of `antlr4::Recognizer`.
+  if (auto* parser = dynamic_cast<antlr4::Parser*>(recognizer)) {
+    AD_CHECK(offendingToken != nullptr);
+    return generateMetadata(parser, offendingToken, line, charPositionInLine);
+  } else if (auto* lexer = dynamic_cast<antlr4::Lexer*>(recognizer)) {
+    // If the recognizer is a Lexer this means that the error was a Lexer error.
+    // In that case the tokens are not yet available and `offendingToken` is
+    // NULL.
```

Comment:
```suggestion
    // In that case  `offendingToken` is `nullptr`.
```

(the `tokens` interface is not visible anymore in this function).